crates/net/discv5/src/config.rs

```diff
@@ -0,0 +1,337 @@
+            }
+        }
+        self
+    }
+
+    /// Set [`ForkId`], and key used to identify it, to set in local [`Enr`](discv5::enr::Enr).
+    pub fn fork(mut self, key: &'static [u8], value: ForkId) -> Self {
+        self.fork = Some((key, value));
+        self
+    }
+
+    /// Sets the tcp port to advertise in the local [`Enr`](discv5::enr::Enr).
+    pub fn tcp_port(mut self, port: u16) -> Self {
+        self.tcp_port = Some(port);
+        self
+    }
+
+    /// Adds an additional kv-pair to include in the local [`Enr`](discv5::enr::Enr).
+    pub fn add_enr_kv_pair(mut self, kv_pair: (&'static str, Bytes)) -> Self {
+        self.other_enr_data.push(kv_pair);
+        self
+    }
+
+    /// Allows discovered nodes without tcp port set in their ENR to be passed from
+    /// [`discv5::Discv5`] up to the app.
+    pub fn allow_no_tcp_discovered_nodes(mut self) -> Self {
+        self.allow_no_tcp_discovered_nodes = true;
+        self
+    }
+
+    /// Adds filter rules to apply to discovered peer to determine whether or not it should be
+    /// passed to the mempool.
+    pub fn filter<F>(self, f: F) -> ConfigBuilder<F>
+    where
+        F: FilterDiscovered,
+    {
+        let Self {
+            discv5_config,
+            bootstrap_nodes,
+            fork,
+            tcp_port,
+            other_enr_data,
+            allow_no_tcp_discovered_nodes,
+            self_lookup_interval,
+            ..
+        } = self;
+
+        ConfigBuilder {
+            discv5_config,
+            bootstrap_nodes,
+            fork,
+            tcp_port,
+            other_enr_data,
+            allow_no_tcp_discovered_nodes,
+            self_lookup_interval,
+            filter_discovered_peer: f,
+        }
+    }
+
+    /// Returns a new [`Config`].
+    pub fn build(self) -> Config<T> {
+        let Self {
+            discv5_config,
+            bootstrap_nodes,
+            fork,
+            tcp_port,
+            other_enr_data,
+            allow_no_tcp_discovered_nodes,
+            self_lookup_interval,
+            filter_discovered_peer,
+        } = self;
+
+        let discv5_config = discv5_config
+            .unwrap_or_else(|| discv5::ConfigBuilder::new(ListenConfig::default()).build());
+
+        let fork = fork.unwrap_or((ChainRef::ETH, MAINNET.latest_fork_id()));
+
+        let tcp_port = tcp_port.unwrap_or(DEFAULT_DISCOVERY_PORT);
+
+        let self_lookup_interval =
+            self_lookup_interval.unwrap_or(DEFAULT_SECONDS_SELF_LOOKUP_INTERVAL);
+
+        Config {
+            discv5_config,
+            bootstrap_nodes,
+            fork,
+            tcp_port,
+            other_enr_data,
+            allow_no_tcp_discovered_nodes,
+            self_lookup_interval,
+            filter_discovered_peer,
+        }
+    }
+}
+
+/// Config used to bootstrap [`discv5::Discv5`].
+#[derive(Debug)]
+pub struct Config<Filter = MustNotIncludeChains> {
```

Comment:
instead of making this generic, let's just include by default a hashset containing the chains that will be disallowed. If we need a "must include" set, this can be added as another hashset (with mutual exclusion checks when building the config) if the feature is really needed